org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/wizard/template/AbstractProjectTemplate.java

```diff
@@ -0,0 +1,122 @@
+/*******************************************************************************
+ * Copyright (c) 2017 itemis AG (http://www.itemis.de) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtext.ui.wizard.template;
+
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import org.apache.log4j.Logger;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Status;
+
+/**
+ * A template definition for a project, used by the new project wizard. Defines the UI (label, description, icon, variables) on how to
+ * present the template to the user and a list or {@link ProjectFile} elements to be generated by the wizard when the template is selected.
+ * The {@link ProjectFile}'s can contain references to {@link ProjectVariable}'s to allow the user to configure the generated code.
+ * 
+ * Instances of this class are generated from classes annotated with {@link ProjectTemplate}.
+ * 
+ * @author Arne Deutsch - Initial contribution and API
+ * @since 2.14
+ */
+public abstract class AbstractProjectTemplate {
+
+	private static final Logger logger = Logger.getLogger(AbstractProjectTemplate.class);
+
+	private List<ProjectVariable> variables;
+
+	protected final StringProjectVariable text(String name, String defaultValue, String description) {
```

Comment:
I do not see any reason for that. The wizard pages know all the subclasses directly to provide apropriate widgets. I have designed the entire API to be just used, not to be extended (e.g. "ProjectVariable" is package protected and all subclasses are final). Enable the user to override these methods would imply that all the other restrictions should be droped also. I would start up with a more closed API and open it if it is requested. Such that we have more oportunities to evolve the API without breaking user code

Reply:
Willl think about your points a little more ...

Reply:
Your points are valid. I will make these methods available for overriding anbd move the widget logic from the parameter page into the variable implementations. This will enable a user to provide his own variables/widgets as needed. Also results in cleaner code and gets rid of the "Pair".